play-kml-tour/src/main/java/com/esri/arcgismaps/sample/playkmltour/MainActivity.kt

```diff
@@ -0,0 +1,252 @@
+import com.google.android.material.snackbar.Snackbar
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import java.io.File
+import kotlin.math.roundToInt
+
+
+class MainActivity : AppCompatActivity() {
+
+    private val TAG = MainActivity::class.java.simpleName
+
+    private val provisionPath: String by lazy {
+        getExternalFilesDir(null)?.path.toString() + File.separator + getString(R.string.app_name)
+    }
+
+    // set up data binding for the activity
+    private val activityMainBinding: ActivityMainBinding by lazy {
+        DataBindingUtil.setContentView(this, R.layout.activity_main)
+    }
+
+    private val sceneView by lazy {
+        activityMainBinding.sceneView
+    }
+
+    private val playPauseButton by lazy {
+        activityMainBinding.playPauseButton
+    }
+
+    private val resetTourButton by lazy {
+        activityMainBinding.resetTourButton
+    }
+
+    private val tourStatusTV by lazy {
+        activityMainBinding.tourStatusTV
+    }
+
+    private val tourProgressBar by lazy {
+        activityMainBinding.tourProgressBar
+    }
+
+    private val kmlTourController = KmlTourController()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        // authentication with an API key or named user is
+        // required to access basemaps and other location services
+        ArcGISEnvironment.apiKey = ApiKey.create(BuildConfig.API_KEY)
+        lifecycle.addObserver(sceneView)
+
+
+        // add elevation data
+        val surface = Surface().apply {
+            elevationSources.add(ArcGISTiledElevationSource(getString(R.string.world_terrain_service)))
+        }
+
+        // create a scene and set the surface
+        sceneView.scene = ArcGISScene(BasemapStyle.ArcGISImagery).apply {
+            baseSurface = surface
+        }
+
+        // add a KML layer from a KML dataset with a KML tour
+        val kmlDataset = KmlDataset(provisionPath + getString(R.string.kml_tour_path))
+        val kmlLayer = KmlLayer(kmlDataset)
+
+        // add the layer to the scene view's operational layers
+        sceneView.scene?.operationalLayers?.add(kmlLayer)
+
+        // load the KML layer
+        lifecycleScope.launch {
+            kmlLayer.load().onFailure {
+                showError(it.message.toString())
+            }.onSuccess {
+                // get the first loaded KML tour
+                val kmlTour = findFirstKMLTour(kmlDataset.rootNodes)
+                if (kmlTour == null) {
+                    showError("Cannot find KML tour in dataset")
+                    return@onSuccess
+                }
+
+                // collect changes in KML tour status
+                collectKmlTourStatus(kmlTour)
+
+                // set the KML tour to the controller
+                kmlTourController.tour = kmlTour
+            }
+        }
+
+        resetTourButton.setOnClickListener { kmlTourController.reset() }
```

Comment:
Could you also reset the KmlLayer be the initial viewpoint?

Reply:
Is there a way to get the initial viewpoint of a KML tour? I tried using `kmlTourController.tour?.viewpoint` but it's always `null`.

Reply:
one way you could do it is when the `play` is clicked, you save the current viewpoint (if it is null), then `reset` is clicked here, then applying the viewpoint here.

Reply:
Added a `initialViewpoint` instance to set `SceneView.setViewpoint` when the reset button is clicked